src/main/java/bio/terra/folder/service/create/CreateService.java

```diff
@@ -0,0 +1,54 @@
+package bio.terra.folder.service.create;
+
+import bio.terra.folder.db.FolderDao;
+import bio.terra.folder.generated.model.CreateFolderBody;
+import bio.terra.folder.service.create.exception.NameConflictException;
+import bio.terra.folder.service.create.flight.FolderCreateFlight;
+import bio.terra.folder.service.create.flight.FolderFlightMapKeys;
+import bio.terra.folder.service.iam.AuthenticatedUserRequest;
+import bio.terra.folder.service.job.JobBuilder;
+import bio.terra.folder.service.job.JobService;
+import java.util.UUID;
+import org.springframework.stereotype.Component;
+
+@Component
+public class CreateService {
+
+  private JobService jobService;
+  private FolderDao folderDao;
+
+  public CreateService(JobService jobService, FolderDao folderDao) {
+    this.jobService = jobService;
+    this.folderDao = folderDao;
+  }
+
+  public String createFolder(CreateFolderBody folderBody, AuthenticatedUserRequest userReq) {
+    String folderId = UUID.randomUUID().toString();
+    String folderName = folderBody.getName();
+    String parentFolderId = folderBody.getParentFolderId().orElse(null);
+
+    if (parentFolderId != null && folderDao.containedFolderNameExists(parentFolderId, folderName)) {
+      throw new NameConflictException(
+          "Folder with name " + folderName + " already exists inside folder " + parentFolderId);
+    }
+    if (parentFolderId != null && folderDao.containedObjectNameExists(parentFolderId, folderName)) {
+      throw new NameConflictException(
+          "Contained object with name "
+              + folderName
+              + " already exists inside folder "
+              + parentFolderId);
+    }
```

Comment:
hopping on Matt's comment about name validation, you could break out a helper method that does all the validation work:
* check syntax of folder name (some regex, probably)
* check length of folder name doesn't overflow db column
* check if folder name exists
* check if contained object of same name exists

also,  these should probably throw a 400, or maybe a 409, to the end user ... which would of course want new tests too.

Reply:
Agreed, and added this in + tests.
The name syntax rules are a little arbitrary, but it seems reasonable to enforce some basic limits around alphanumeric/spaces/special characters etc. I'll update the design doc with this as well.